File: app/src/main/java/org/deiverbum/app/data/entity/relation/TodayCompletas.java
package org.deiverbum.app.data.entity.relation;

import androidx.room.Embedded;
import androidx.room.Relation;

import org.deiverbum.app.data.entity.LHHymnJoinEntity;
import org.deiverbum.app.data.entity.LHPrayerEntity;
import org.deiverbum.app.data.entity.LHPsalmodyEntity;
import org.deiverbum.app.data.entity.LHPsalmodyJoinEntity;
import org.deiverbum.app.data.entity.LHReadingShortJoinEntity;
import org.deiverbum.app.data.entity.LiturgyEntity;
import org.deiverbum.app.data.entity.TodayEntity;
import org.deiverbum.app.model.BiblicalShort;
import org.deiverbum.app.model.LHHymn;
import org.deiverbum.app.model.LHPsalmody;
import org.deiverbum.app.model.Liturgy;
import org.deiverbum.app.model.Prayer;
import org.deiverbum.app.model.Today;

import java.util.List;

/**
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1
 */
public class TodayCompletas {

    @Embedded
    public TodayEntity today;


    @Relation(
            entity = LHHymnJoinEntity.class,
            parentColumn = "tHymnFK",
            entityColumn = "groupID"
    )
    public LHHymnWithAll himno;

    @Relation(
            entity = LHReadingShortJoinEntity.class,
            parentColumn = "tBiblicalFK",
            entityColumn = "groupID"
    )
    public LHReadingShortAll biblica;

    @Relation(
            entity = LHPsalmodyJoinEntity.class,
            parentColumn = "tPsalmodyFK",
            entityColumn = "groupID"
    )
    public LHPsalmodyAll salmodia;

    @Relation(
            entity = LHPsalmodyEntity.class,
            parentColumn = "tPsalmodyFK",
            entityColumn = "groupFK"
    )
    public List<PsalmodyWithPsalms> salmos;

    @Relation(
            entity = LHPrayerEntity.class,
            parentColumn = "tPrayerFK",
            entityColumn = "groupID"
    )
    public LHPrayerAll lhPrayerAll;

    @Relation(
            entity = LiturgyEntity.class,
            parentColumn = "liturgyFK",
            entityColumn = "liturgyID"
    )
    public LiturgyWithTime feria;

    @Relation(
            entity = LiturgyEntity.class,
            parentColumn = "previousFK",
            entityColumn = "liturgyID"
    )
    public LiturgyWithTime previo;

    public Today getToday() {
        Today dm = new Today();
        int wd = today.weekDay;
        dm.setWeekDay(wd - 1);
        dm.liturgyDay = feria.getDomainModel();
        dm.liturgyPrevious = today.previoId > 1 ? previo.getDomainModel() : null;
        dm.setTimeID(today.tiempoId);
        dm.setTodayDate(today.getHoy());
        return dm;
    }

    public Liturgy getDomainModel() {
        Liturgy dm = feria.getDomainModel();
        dm.typeID = 7;
        dm.setToday(getToday());
        return dm;
    }

    public LHHymn getHimno() {
        return himno.getDomainModel();
    }

    public BiblicalShort getBiblica() {
        return biblica.getDomainModel(today.getTiempoId());
    }

    public LHPsalmody getSalmodia() {
        return salmodia.getDomainModel();
    }

    public Prayer getOracion() {
        return lhPrayerAll.getDomainModel();
    }

}

File: app/src/main/java/org/deiverbum/app/data/entity/relation/TodayCompletas.kt
package org.deiverbum.app.data.entity.relation

import androidx.room.Embedded
import androidx.room.Relation
import org.deiverbum.app.data.entity.LHNightPrayerEntity
import org.deiverbum.app.data.entity.LiturgyEntity
import org.deiverbum.app.data.entity.TodayEntity
import org.deiverbum.app.model.BreviaryHour
import org.deiverbum.app.model.Completas
import org.deiverbum.app.model.Conclusion
import org.deiverbum.app.model.RitosIniciales
import org.deiverbum.app.model.Today

/**
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1
 */
data class TodayCompletas(
    @Embedded
    var today: TodayEntity,

    @Relation(
        entity = LHNightPrayerEntity::class,
        parentColumn = "nightPrayerFK",
        entityColumn = "groupID"
    )
    var nightPrayer: NightPrayerWithAll,

    @Relation(entity = LiturgyEntity::class, parentColumn = "liturgyFK", entityColumn = "liturgyID")
    var liturgy: LiturgyWithTime
) {

    val domainModel: Today
        get() {
            val dmToday = Today()
            dmToday.saintFK = today.saintFK
            dmToday.liturgyDay = liturgy.domainModel
            dmToday.todayDate = today.todayDate
            dmToday.hasSaint = today.hasSaint
            dmToday.liturgyDay.typeID = 7
            val bh = BreviaryHour()
            val completas = Completas()
            val ritosIniciales = RitosIniciales()
            val kyrie = nightPrayer.kyrie!!.kyrieEntity!!.getDomainModel()
            //kyrie.setIntroduccion("Es muy de alabar que, después de la invocación inicial, se haga el examen de conciencia, el cual en la celebración comunitaria puede concluirse con un acto penitencial, de la siguiente forma:|Hermanos, habiendo llegado al final de esta jornada que Dios nos ha concedido, reconozcamos sinceramente nuestros pecados.|Todos examinan en silencio su conciencia. Terminando el examen se añade la fórmula penitencial:")
            //kyrie.setConclusion("Pueden usarse otras invocaciones penitenciales. <br />Si preside la celebración un ministro, él solo dice la absolución siguiente; en caso de lo contrario la dicen todos: <br />|El Señor todopoderoso tenga misericordia de nosotros, perdone nuestros pecados y nos lleve a la vida eterna.∬Amén.")
            ritosIniciales.kyrie = kyrie
            completas.setRitosIniciales(ritosIniciales)
            completas.setHimno(nightPrayer.hymn!!.domainModel)
            completas.setSalmodia(nightPrayer.salmodia!!.domainModel)
            completas.biblicalShort = nightPrayer.readingFK!!.getDomainModel(today.timeID)
            completas.setNuncDimitis(nightPrayer.nuncDimitisFK!!.getDomainModel(today.timeID))
            completas.setOracion(nightPrayer.prayer!!.domainModel)
            completas.setConclusion(getConclusion())
            bh.setCompletas(completas)
            dmToday.liturgyDay.breviaryHour = bh
            return dmToday
        }

    private fun getConclusion(): Conclusion {
        val conclusion = Conclusion()
        conclusion.setAntifonaVirgen(nightPrayer.virgin!!.virginEntity!!.antiphon)
        return conclusion
    }
}